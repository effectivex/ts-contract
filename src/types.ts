// tslint:disable:no-shadowed-variable
import * as Logger from 'bunyan';
import * as Joi from './joi';

export type ConvertType3<T> = T extends JoiPrimitiveSchema ? ExtractPrimitive<T> : T;

export type ConvertType2<T> = T extends JoiPrimitiveSchema
  ? ExtractPrimitive<T>
  : T extends Joi.ArraySchema<infer K>
    ? Array<K extends Joi.ObjectSchema<infer P> ? ExtractObject3<P> : ConvertType3<K>>
    : T extends Joi.ObjectSchema<infer K> ? ExtractObject3<K> : T;

export type ExtractObject2<T> = { [K in keyof T]: ConvertType2<T[K]> };
export type ExtractObject3<T> = { [K in keyof T]: ConvertType3<T[K]> };

export type JoiPrimitiveSchema =
  | Joi.StringSchema
  | Joi.NumberSchema
  | Joi.BooleanSchema
  | Joi.DateSchema;

export type ExtractPrimitive<T> = T extends Joi.StringSchema<infer REQ>
  ? (REQ extends true ? string : (string | undefined))
  : T extends Joi.NumberSchema<infer REQ>
    ? (REQ extends true ? number : (number | undefined))
    : T extends Joi.BooleanSchema ? boolean : T extends Joi.DateSchema ? Date : T;

export type ConvertType<T> = T extends JoiPrimitiveSchema
  ? ExtractPrimitive<T>
  : T extends Joi.ArraySchema<infer K>
    ? Array<
        K extends Joi.ObjectSchema<infer P, infer REQ>
          ? (REQ extends true ? ExtractObject2<P> : (ExtractObject2<P> | undefined))
          : ConvertType2<K>
      >
    : T extends Joi.ObjectSchema<infer K, infer REQ>
      ? (REQ extends true ? ExtractObject2<K> : (ExtractObject2<K> | undefined))
      : T;

export type ExtractObject<T> = { [K in keyof T]: ConvertType<T[K]> };

export interface ContractOptions {
  sync: boolean;
  removeOutput: boolean;
}

export interface ContractConfig {
  removeFields: string[];
  debug: boolean;
  depth: number;
  maxArrayLength: number;
  getLogger: (serviceName: string) => Logger;
  getNextId: () => number;
}

export interface Contract {
  <
    ARG1 extends string,
    ARG2 extends string,
    ARG3 extends string,
    ARG4 extends string,
    T extends { [key in ARG1 | ARG2 | ARG3 | ARG4]: Joi.SchemaLike },
    R,
    FN extends (
      arg1: ConvertType<T[ARG1]>,
      arg2: ConvertType<T[ARG2]>,
      arg3: ConvertType<T[ARG3]>,
      arg4: ConvertType<T[ARG4]>,
    ) => R
  >(
    signature: string,
    params: [ARG1, ARG2, ARG3, ARG4],
    schema: T,
    fn: FN,
    options?: Partial<ContractOptions>,
  ): FN;

  <
    ARG1 extends string,
    ARG2 extends string,
    ARG3 extends string,
    T extends { [key in ARG1 | ARG2 | ARG3]: Joi.SchemaLike },
    R,
    FN extends (
      arg1: ConvertType<T[ARG1]>,
      arg2: ConvertType<T[ARG2]>,
      arg3: ConvertType<T[ARG3]>,
    ) => R
  >(
    signature: string,
    params: [ARG1, ARG2, ARG3],
    schema: T,
    fn: FN,
    options?: Partial<ContractOptions>,
  ): FN;

  <
    ARG1 extends string,
    ARG2 extends string,
    T extends { [key in ARG1 | ARG2]: Joi.SchemaLike },
    R,
    FN extends (arg1: ConvertType<T[ARG1]>, arg2: ConvertType<T[ARG2]>) => R
  >(
    signature: string,
    params: [ARG1, ARG2],
    schema: T,
    fn: FN,
    options?: Partial<ContractOptions>,
  ): FN;

  <
    ARG1 extends string,
    T extends { [key in ARG1]: Joi.SchemaLike },
    R,
    FN extends (arg1: ConvertType<T[ARG1]>) => R
  >(
    signature: string,
    params: [ARG1],
    schema: T,
    fn: FN,
    options?: Partial<ContractOptions>,
  ): FN;

  <T extends {}, R, FN extends () => R>(
    signature: string,
    params: undefined[],
    schema: T,
    fn: FN,
    options?: Partial<ContractOptions>,
  ): FN;
}
