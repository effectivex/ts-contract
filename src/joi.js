// export * from 'joi';
const joi = require('joi');

module.exports = joi;
