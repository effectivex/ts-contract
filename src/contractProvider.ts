import * as Logger from 'bunyan';
import * as Joi from './joi';
import { ConvertType, ContractOptions, ContractConfig, Contract } from './types';
import { wrapValidate } from './wrapValidate';
import { wrapLog } from './wrapLog';

const getConfig = (config?: Partial<ContractConfig>): ContractConfig => {
  let defaultId = 0;

  return {
    removeFields: ['password', 'token', 'accessToken'],
    debug: true,
    depth: 4,
    maxArrayLength: 30,
    getLogger(serviceName: string) {
      return Logger.createLogger({
        name: serviceName,
        level: this.debug ? 'debug' : 'error',
      });
    },
    getNextId: () => ++defaultId,
    ...(config || {}),
  };
};

const getOptions = (options?: Partial<ContractOptions>): ContractOptions => {
  return {
    sync: false,
    removeOutput: false,
    ...(options || {}),
  };
};

/**
 * Create a new contract wrapper
 * @param overrideConfig the config
 */
export const contractProvider = (overrideConfig?: Partial<ContractConfig>): Contract => {
  const config = getConfig(overrideConfig);
  // function contract<
  //   ARG1 extends string,
  //   ARG2 extends string,
  //   T extends { [key in ARG1 | ARG2]: Joi.SchemaLike },
  //   R,
  //   FN extends (arg1: ConvertType<T[ARG1]>, arg2: ConvertType<T[ARG2]>) => R
  // >(signature: string, params: [ARG1, ARG2], schema: T, fn: FN, options?: ContractOptions): FN;

  // function contract<
  //   ARG1 extends string,
  //   T extends { [key in ARG1]: Joi.SchemaLike },
  //   R,
  //   FN extends (arg1: ConvertType<T[ARG1]>) => R
  // >(signature: string, params: [ARG1], schema: T, fn: FN, options?: ContractOptions): FN;

  function contract(
    signature: string,
    params: string[],
    schema: Record<string, Joi.SchemaLike>,
    fn: () => any,
    overrideOptions?: Partial<ContractOptions>,
  ): any {
    const [serviceName, methodName] = signature.split('#');
    if (!methodName) {
      throw new Error(
        `Invalid signature. Must be in format "service-name#method-name". Received "${signature}".`,
      );
    }
    return function decorated(this: any, ...args: any[]) {
      const options = getOptions(overrideOptions);
      const logger = config.getLogger(serviceName);
      const withValidation = wrapValidate({
        keysSchema: schema,
        method: fn,
        paramNames: params,
        sync: options.sync,
      });

      const withLogging = wrapLog({
        logger,
        method: withValidation.bind(this),
        methodName,
        paramNames: params,
        config,
        removeOutput: options.removeOutput,
      });
      return withLogging.call(this, ...args);
    };
  }

  return contract;
};
