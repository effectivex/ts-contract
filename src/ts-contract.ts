import * as Joi from './joi';

export * from './types';
export * from './contractProvider';
export { Joi };
