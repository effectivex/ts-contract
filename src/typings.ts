import * as Joi from './joi';
import { ConvertType, ExtractObject, ExtractObjectRequired } from './types';

const schema = {
  optionalString: Joi.string(),
  requiredString: Joi.string().required(),
};
const a = Joi.string();
type A = ConvertType<typeof a>;
type StringData = ExtractObject<typeof schema>;
const stringData: StringData = {
  requiredString: 'str',
  // optionalString: undefined,
};

type NonNeverNames<T> = { [K in keyof T]: T[K] extends (null | undefined) ? never : K }[keyof T];

type BData = {
  a: string;
  b: string | undefined;
  c: never;
};

const bdata: BData = {
  a: 'str',
  b: 'str',
};

type ValueOf<T> = T[keyof T];

type B = NonNeverNames<BData>;

type FilterNever<T> = Pick<T, NonNeverNames<T>>;

type C = FilterNever<BData>;

// type C = NonNullable<BData>;
// type D = keyof BData;
// type E = ValueOf<BData>;
// type F = Pick<BData, 'a' | 'b'>;

// type GetRequired<T> = { [P in NonNullable<keyof T>]: NonNullable<T[P]> };

// type G = GetRequired<BData>;
