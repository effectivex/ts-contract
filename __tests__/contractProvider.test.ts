import * as Logger from 'bunyan';
import * as Joi from '../src/joi';
import { Contract } from '../src/types';
import { contractProvider } from '../src/contractProvider';

let logger: Logger;
let contract: Contract;

beforeEach(() => {
  logger = getLogger();
  contract = contractProvider({
    getLogger: () => logger,
  });
});

function getLogger() {
  return ({
    error: jest.fn(),
    debug: jest.fn(),
  } as any) as Logger;
}

it('should throw if signature is invalid', () => {
  expect(() => {
    contract('foo', [], {}, () => 10, { sync: true });
  }).toThrowErrorMatchingSnapshot();
});

it('validate sync function', () => {
  const fn = contract(
    'myService#fn',
    ['a'],
    {
      a: Joi.number()
        .positive()
        .required(),
    },
    a => a + 10,
    { sync: true },
  );
  expect(() => {
    fn(10);
  }).not.toThrow();
  expect(() => {
    fn(-10);
  }).toThrowErrorMatchingSnapshot();
  expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
  expect((logger.error as jest.Mock<any>).mock.calls).toMatchSnapshot();
});

it('validate sync with nested object', () => {
  const fn = contract(
    'myService#fn',
    ['params'],
    {
      params: Joi.object({
        username: Joi.string()
          .min(2)
          .required(),
        inner: Joi.object({
          b: Joi.number()
            .min(0)
            .required(),
        }).required(),
      }).required(),
    },
    params => params.username + '-' + params.inner.b,
    { sync: true },
  );
  expect(() => {
    fn({ username: 'baz', inner: { b: -1 } });
  }).toThrowErrorMatchingSnapshot();
  expect(() => {
    fn({ username: 'b', inner: { b: 10 } });
  }).toThrowErrorMatchingSnapshot();
  expect(fn({ username: 'bar', inner: { b: 10 } })).toEqual('bar-10');
});

it('validate async function', async () => {
  const fn = contract(
    'myService#fn',
    ['a'],
    {
      a: Joi.number()
        .positive()
        .required(),
    },
    async a => a + 10,
  );
  await expect(fn(10)).resolves.toBe(20);
  await expect(fn(-10)).rejects.toThrowErrorMatchingSnapshot();
  expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
  expect((logger.error as jest.Mock<any>).mock.calls).toMatchSnapshot();
});

it('with removeOutput', () => {
  const fn = contract(
    'myService#fn',
    ['a'],
    {
      a: Joi.number()
        .positive()
        .required(),
    },
    a => a + 10,
    { sync: true, removeOutput: true },
  );
  fn(10);
  expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
});

it('should keep `this` context', () => {
  class MyService {
    fn = contract(
      'myService#fn',
      ['a'],
      {
        a: Joi.number()
          .positive()
          .required(),
      },
      a => a + this.getInt(),
      { sync: true, removeOutput: true },
    );
    private getInt() {
      return 10;
    }
  }
  const myService = new MyService();
  expect(myService.fn(5)).toBe(15);
});
