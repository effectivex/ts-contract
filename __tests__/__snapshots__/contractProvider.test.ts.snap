// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should throw if signature is invalid 1`] = `"Invalid signature. Must be in format \\"service-name#method-name\\". Received \\"foo\\"."`;

exports[`validate async function 1`] = `
"{
  \\"a\\" [31m[1][0m: -10
}
[31m
[1] \\"a\\" must be a positive number[0m"
`;

exports[`validate async function 2`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER fn:",
    "{ a: 10 }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT fn:",
    "20",
  ],
  Array [
    Object {
      "id": 2,
    },
    "ENTER fn:",
    "{ a: -10 }",
  ],
]
`;

exports[`validate async function 3`] = `
Array [
  Array [
    Object {
      "id": 2,
    },
    "ERROR fn: { a: -10 } 
",
    [ValidationError: {
  "a" [31m[1][0m: -10
}
[31m
[1] "a" must be a positive number[0m],
  ],
]
`;

exports[`validate sync function 1`] = `
"{
  \\"a\\" [31m[1][0m: -10
}
[31m
[1] \\"a\\" must be a positive number[0m"
`;

exports[`validate sync function 2`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER fn:",
    "{ a: 10 }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT fn:",
    "20",
  ],
  Array [
    Object {
      "id": 2,
    },
    "ENTER fn:",
    "{ a: -10 }",
  ],
]
`;

exports[`validate sync function 3`] = `
Array [
  Array [
    Object {
      "id": 2,
    },
    "ERROR fn: { a: -10 } 
",
    [ValidationError: {
  "a" [31m[1][0m: -10
}
[31m
[1] "a" must be a positive number[0m],
  ],
]
`;

exports[`validate sync with nested object 1`] = `
"{
  \\"params\\": {
    \\"username\\": \\"baz\\",
    \\"inner\\": {
      \\"b\\" [31m[1][0m: -1
    }
  }
}
[31m
[1] \\"b\\" must be larger than or equal to 0[0m"
`;

exports[`validate sync with nested object 2`] = `
"{
  \\"params\\": {
    \\"inner\\": {
      \\"b\\": 10
    },
    \\"username\\" [31m[1][0m: \\"b\\"
  }
}
[31m
[1] \\"username\\" length must be at least 2 characters long[0m"
`;

exports[`with removeOutput 1`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER fn:",
    "{ a: 10 }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT fn:",
    "<removed>",
  ],
]
`;
